### PR TITLE
scheduler: refactor main entry Run()

### PR DESCRIPTION
**Pull request by hongchaodeng**
The kube-scheduler/app.Run() is the main entry of scheduler program.
It's enormous.
This PR tries to clean it up. Should be more modular and readable.

### CODE REVIEW

**Comment by hongchaodeng** [comment]
cc @timothysc @xiang90 
Can you help review this?

**Comment by hongchaodeng** [comment]
@k8s-bot cri e2e test this

**Comment by hongchaodeng** [comment]
@k8s-bot cri e2e test this

**Comment by hongchaodeng** [comment]
@k8s-bot cri e2e test this.

File: plugin/cmd/kube-scheduler/scheduler.go [hunk 1]
```diff
@@ -36,5 +37,8 @@ func main() {
 
 	verflag.PrintAndExitIfRequested()
 
-	app.Run(s)
+	if err := app.Run(s); err != nil {
+		glog.Errorf("scheduler app failed to run: %v", err)
```
Comment (xiang90) [inline plugin/cmd/kube-scheduler/scheduler.go]: glog.Fatal here? or we can reach line 43.

File: plugin/cmd/kube-scheduler/scheduler.go [hunk 2]
```diff
@@ -36,5 +37,8 @@ func main() {
...
+	}
+	glog.Fatal("this statement shouldn't be reached")
```
Comment (xiang90) [inline plugin/cmd/kube-scheduler/scheduler.go]: we should panic on a unreachable line not fatal. usually fatal on unrecoverable external error, and panic on unrecoverable internal error.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 3]
```diff
@@ -73,97 +73,46 @@ through the API as necessary.`,
...
 	run := func(_ <-chan struct{}) {
 		sched.Run()
 		select {}
 	}
 
 	if !s.LeaderElection.LeaderElect {
 		run(nil)
-		glog.Fatal("this statement is unreachable")
-		panic("unreachable")
+		return nil
```
Comment (xiang90) [inline plugin/cmd/kube-scheduler/app/server.go]: probably should still panic internally.

File: plugin/cmd/kube-scheduler/app/server.go [hunk 4]
```diff
@@ -174,12 +123,56 @@ func Run(s *options.SchedulerServer) error {
...
+	}
+
+	kubeconfig.ContentType = s.ContentType
+	// Override kubeconfig qps/burst settings from flags
+	kubeconfig.QPS = s.KubeAPIQPS
+	kubeconfig.Burst = int(s.KubeAPIBurst)
+
+	cli, err := clientset.NewForConfig(restclient.AddUserAgent(kubeconfig, "leader-election"))
+	if err != nil {
+		return nil, fmt.Errorf("Invalid API configuration: %v", err)
```
Comment (xiang90) [inline plugin/cmd/kube-scheduler/app/server.go]: invalid API configuration: %v

File: plugin/cmd/kube-scheduler/app/server.go [hunk 5]
```diff
@@ -174,12 +123,56 @@ func Run(s *options.SchedulerServer) error {
...
+		Handler: mux,
+	}
+	glog.Fatal(server.ListenAndServe())
 }
 
-func createConfig(s *options.SchedulerServer, configFactory *factory.ConfigFactory) (*scheduler.Config, error) {
+func createClient(s *options.SchedulerServer) (*clientset.Clientset, error) {
+	kubeconfig, err := clientcmd.BuildConfigFromFlags(s.Master, s.Kubeconfig)
+	if err != nil {
+		glog.Errorf("unable to build config from flags: %v", err)
```
Comment (xiang90) [inline plugin/cmd/kube-scheduler/app/server.go]: we should log error in the caller?

**Comment by xiang90** [comment]
Minor issues. LGTM after fixing them.

**Comment by hongchaodeng** [comment]
Fixed minor issues

**Comment by xiang90** [comment]
@hongchaodeng Awesome. Thanks!

**Comment by hongchaodeng** [comment]
@k8s-bot unit test this #38594

**Comment by hongchaodeng** [comment]
@k8s-bot cri e2e test this #38556

**Comment by hongchaodeng** [comment]
@k8s-bot cri e2e test this

**Comment by hongchaodeng** [comment]
@wojtek

Can you help review?

File: plugin/cmd/kube-scheduler/app/server.go [hunk 6]
```diff
@@ -174,12 +123,55 @@ func Run(s *options.SchedulerServer) error {
...
+	healthz.InstallHandler(mux)
+	if s.EnableProfiling {
+		mux.HandleFunc("/debug/pprof/", pprof.Index)
+		mux.HandleFunc("/debug/pprof/profile", pprof.Profile)
+		mux.HandleFunc("/debug/pprof/symbol", pprof.Symbol)
+		if s.EnableContentionProfiling {
+			goruntime.SetBlockProfileRate(1)
+		}
+	}
+	if c, err := configz.New("componentconfig"); err == nil {
```
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/server.go]: It's difficult to tell, but did you ensure this was the 1st place configz was used in the shuffling.  B4 it was the 1st line before refactoring.

File: plugin/cmd/kube-scheduler/scheduler.go [hunk 7]
```diff
@@ -36,5 +37,8 @@ func main() {
 
 	verflag.PrintAndExitIfRequested()
 
-	app.Run(s)
+	if err := app.Run(s); err != nil {
+		glog.Fatalf("scheduler app failed to run: %v", err)
+	}
+	panic("this statement shouldn't be reached")
```
Comment (timothysc) [inline plugin/cmd/kube-scheduler/scheduler.go]: could we change this panic to be "unreachable" to match the rest of the codebase?  It's the only outlier.

Comment (timothysc) [inline plugin/cmd/kube-scheduler/scheduler.go]: I don't think this logic makes much sense.. of course we can reach this point via a number of ways.  Shouldn't we just exit gracefully?

File: plugin/cmd/kube-scheduler/app/server.go [hunk 8]
```diff
@@ -73,97 +73,46 @@ through the API as necessary.`,
...
-
-	kubeconfig.ContentType = s.ContentType
-	// Override kubeconfig qps/burst settings from flags
-	kubeconfig.QPS = s.KubeAPIQPS
-	kubeconfig.Burst = int(s.KubeAPIBurst)
-
+	kubecli, err := createClient(s)
 	if err != nil {
-		glog.Fatalf("Invalid API configuration: %v", err)
+		return fmt.Errorf("unable to create kube client: %v", err)
```
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/server.go]: Does it make sense to re-wrap the error or just return them?   I just debate the utility if we're refactoring.

**Review by timothysc — CHANGES_REQUESTED** [review]
Please answer the minor comments, otherwise lgtm

File: plugin/cmd/kube-scheduler/app/server.go (hunk 8 above)
Comment (hongchaodeng) [inline plugin/cmd/kube-scheduler/app/server.go]: The previous code usually Fatalf() right there. Now We Fatalf() outside, and thus it's better to wrap them.

File: plugin/cmd/kube-scheduler/scheduler.go (hunk 7 above)
Comment (hongchaodeng) [inline plugin/cmd/kube-scheduler/scheduler.go]: I will remove this one.

> of course we can reach this point via a number of ways.

IIRC, no way to reach here? Can you clarify?

File: plugin/cmd/kube-scheduler/app/server.go (hunk 6 above)
Comment (hongchaodeng) [inline plugin/cmd/kube-scheduler/app/server.go]: Not sure why it needs to be put 1st line.
cc original author @mikedanese

File: plugin/cmd/kube-scheduler/app/server.go (hunk 8 above)
Comment (timothysc) [inline plugin/cmd/kube-scheduler/app/server.go]: k.

**Review by timothysc — APPROVED** [review]

**Comment by timothysc** [comment]
lgtm.

**Pull request merged**